src/Native/Unix/System.Native/pal_networking.cpp

```diff
@@ -2157,12 +2163,12 @@ SystemNative_SetSockOpt(intptr_t socket, int32_t socketOptionLevel, int32_t sock
         }
         else if (socketOptionName == PAL_SO_REUSEADDR)
         {
-            //
-            // On Windows, SO_REUSEADDR allows the address *and* port to be reused.  It's equivalent to 
-            // SO_REUSEADDR + SO_REUSEPORT other systems. 
-            //
+            // SO_REUSEPORT on Linux will do load-balancing accross multiple TCP sockets.
+            // This is different from Windows and BSD-Unix, so we won't set this for TCP.
```

Comment:
Have you validated these behaviors on macOS?

Reply:
Re: "This is different from Windows".  The docs for SO_REUSEADDR state: "Once the second socket has successfully bound, the behavior for all sockets bound to that port is indeterminate. For example, if all of the sockets on the same port provide TCP service, any incoming TCP connection requests over the port cannot be guaranteed to be handled by the correct socket"... so it seems like the Windows behavior is undefined and load-balancing would be an appropriate result.

Reply:
> Have you validated these behaviors on macOS?

Only by the CI tests.

> Re: "This is different from Windows". The docs for SO_REUSEADDR state: "Once the second socket has successfully bound, the behavior for all sockets bound to that port is indeterminate. For example, if all of the sockets on the same port provide TCP service, any incoming TCP connection requests over the port cannot be guaranteed to be handled by the correct socket"... so it seems like the Windows behavior is undefined and load-balancing would be an appropriate result.

On Unix, ReuseAddress controls both TCP_WAIT binding AND socket address sharing.
On Windows, ReuseAddress only controls socket address sharing. TCP_WAIT binding is always allowed.
It's not possible to express on Unix, allow bind during TCP_WAIT without allowing socket address sharing.
On Windows (and BSD Unix) I believe the socket address sharing is used to hand over a socket address from one process to another. So you either get the old one, or the new one.
On Linux, by allowing the socket address sharing, a newly bound process jumps in immediately and starts handing part of the requests, so you get a mix.

So, to align the behavior regarding TCP_WAIT, this PR is setting SO_REUSEADDR on Bind.
To avoid unintentional socket address sharing, it no longer sets SO_REUSEPORT for ReuseAddress.
To have some backwards compatibility it uses SO_REUSEADDR as a 'fake' toggle for ReuseAddress (but it gets forced to 1 on Bind).

I think if we want to add SO_REUSEPORT behavior to Unix, we should do it only when ExclusiveAddressUse=false. This will give explicit control if sharing is desired.